### ISSUE TITLE
Building for MATLAB R2018a fails due to deprecation

### ISSUE THREAD

**Issue by 5ht2** [astra-toolbox/astra-toolbox#149]
During building from source files for MATLAB R2018a I got the following deprecation error:

```Function mxCreateSharedDataCopy not supported.  Either switch to using the more efficient MATLAB Data API available in C++ MEX Applications</a>, or rerun MEX with the -R2017b command line option to use the R2017b API.```

However, this function is only used in a single class (`CFloat32CustomMemoryMatlab3D`), which, in turn, doesn't seem to be used anywhere. Also, this class is wrapped in `#ifdef USE_MATLAB_UNDOCUMENTED` construction, so I was wondering what it was meant for.
For the time being, I just excluded this class from compilation, and everything seems to work fine so far.

**Comment by wjp** [comment]
This class is used for the `astra_mex_data3d_c('link')` functionality, which allows you to directly use a (single precision) matlab array without copying the data.

Thank you for pointing out this deprecation error. For now I would recommend using the `-R2017b` option as suggested in the error. It seems we will have to switch to the new MATLAB Data API for new versions of MATLAB, and keep the existing code for supporting older versions. I will take a look at this when we upgrade to R2018a here.

**Comment by 5ht2** [comment]
For some reason, I couldn't fix it by passing `-R2017b` flag. Furthermore, it turns out that mex compiler should use this flag by default, and I am able to successfully compile a simple test module containing this `mxCreateSharedDataCopy` function - by no luck during Astra build.
Anyway, I'll wait until someone can reproduce this bug.

**Comment by aljalali** [comment]
Hello friends
To install astra toolbox it's possible to guide me.
  I did not get the installation process.
For example, the meaning of
Copy boost headers to lib\include\boost (i.e., copy the boost subdirectory from the boost source archive to lib\include), and boost libraries to lib\x64.
What is?
How to do it using Visual Studio 2017
plz help me

**Comment by wjp** [comment]
@aljalali  Please don't re-use an issue to ask an unrelated question. However, is there a reason you are trying to build from source instead of directly using the binary version?

**Comment by theantisnear** [comment]
I am having the same problem:

Function mxCreateSharedDataCopy not supported.  Either switch to using the more efficient <a href="matlab: helpview([docroot '/matlab/helptargets.map'],'matlab-data-api')">MATLAB Data API</a> available in <a href="matlab: helpview([docroot '/matlab/helptargets.map'],'cpp-mex-applications')">C++ MEX Applications</a>, or rerun MEX with the -R2017b command line option to use the R2017b API.

**Comment by nabriis** [comment]
How did you exclude the class from compilation?

**Comment by 5ht2** [comment]
I just commented out `#define USE_MATLAB_UNDOCUMENTED` line in `mexDataManagerHelpFunctions.cpp` (line 39). If I understood correctly, the side effect of this is that linking functionality in Astra will be disabled.

**Comment by aljalali** [comment]
Hi, I could not install the software, it's possible to take a movie
from the installation phase ... please

۱۳۹۷-۰۶-۱۳ ۱۶:۳۹ ‎+۰۴:۳۰ گرینویچ, 5ht2 <notifications@github.com>:

**Comment by 5ht2** [comment]
@aljalali Could you please create a separate issue with your question?

**Comment by wjp** [comment]
Building for R2017a shouldn't require gcc 4.9. Matlab might throw warnings with newer versions, but as far as I can tell, they are harmless.

I'll try to make some time soon to investigate the issue and either adjust the code or the build system to fix this.

**Comment by wjp** [comment]
Update: it seems the reason that the `-R2017b` flag to mex isn't working with R2018b is that this doesn't affect the main astra code (which isn't compiled using mex). It looks like adding `-DMATLAB_MEXCMD_RELEASE=700` to the main `CPPFLAGS` will work around that.

**Comment by PavelParamonov** [comment]
There is another problem related to this topic. I got similar issue as @askorikov described while building from Visual Studio 2017, but as a linking error:

`error LNK2019: unresolved external symbol mxCreateSharedDataCopyIsDeprecated referenced in function "public: __cdecl CFloat32CustomMemoryMatlab3D::CFloat32CustomMemoryMatlab3D(struct mxArray_tag const *,bool,unsigned __int64)" (??0CFloat32CustomMemoryMatlab3D@@QEAA@PEBUmxArray_tag@@_N_K@Z)`

@wjp, thank you, `-DMATLAB_MEXCMD_RELEASE=700` worked fine! Need to mention, that it looks like `-DMATLAB_DEFAULT_RELEASE=700` does the same job.

However, then another sort of error appeared:

`astra-toolbox-master\matlab\mex\mexdatamanagerhelpfunctions.cpp(42): error C2733: 'mxUnshareArray': second C linkage of overloaded function not allowed
c:\program files\matlab\r2018a\extern\include\matrix.h(1478): note: see declaration of 'mxUnshareArray' (compiling source file mexDataManagerHelpFunctions.cpp)`

It seems that `bool mxUnshareArray(mxArray *pr, bool noDeepCopy);` has been defined in `matrix.h`, but with different signature:

`LIBMMWMATRIX_PUBLISHED_API_EXTERN_C int mxUnshareArray(mxArray *pa, int level);`

So I matched the signature, and it all worked well. @wjp , what do you think of this? Was it correct work around? Thank you for help!

**Comment by wjp** [comment]
Yes, that indeed sounds like the right thing to do for `mxUnshareArray`. Thanks for the feedback!

**Comment by PranjalSahu** [comment]
> Update: it seems the reason that the `-R2017b` flag to mex isn't working with R2018b is that this doesn't affect the main astra code (which isn't compiled using mex). It looks like adding `-DMATLAB_MEXCMD_RELEASE=700` to the main `CPPFLAGS` will work around that.

It does not work for me. The DMATLAB_MEXCMD_RELEASE flag is coming as default in the command executed, however I am still getting the message of using -R2017b flag.
